src/hash.hpp

```diff
@@ -13,115 +15,138 @@
-    template <typename Hash=jenkins>
-    class Simhash {
+    class HasherInterface {
+    public:
+        virtual void update(hash_t) = 0;
+        virtual hash_t result() = 0;
+    };
+
+    class Simhash: public HasherInterface {
     private:
-        class Accumulator {
-        private:
-            static const size_t BITS = sizeof(hash_t) * 8;
-
-            size_t window;
-            Cyclic<hash_t> cyclic;
-            std::vector<size_t> v;
-            size_t hash_count;
-        public:
-            Accumulator() : window(3), cyclic(window), v(BITS, 0), hash_count(0) {}
-
-            void update(hash_t hash) {
-                hash = cyclic.push(hash);
-                ++hash_count;
-                for (int j = BITS - 1; j >= 0; --j) {
-                    v[j] += hash & 1;
-                    hash >>= 1;
-                }
+        static const size_t BITS = sizeof(hash_t) * 8;
+
+        std::vector<size_t> bit_counts;
+        size_t hash_count;
+    public:
+        Simhash(): bit_counts(BITS, 0), hash_count(0) {}
+
+        /**
+         * Updates bit counts with pre-calculated hash.
+         */
+        virtual void update(hash_t hash) /* override */ {
+            ++hash_count;
+            for (int j = BITS - 1; j >= 0; --j) {
+                bit_counts[j] += hash & 1;
+                hash >>= 1;
             }
+        }
 
-            /* With counts appropriately tallied, create a 1 bit for each of
-             * the counts that's over threshold. That result is the hash. */
-            hash_t result() {
-                size_t threshold = hash_count / 2;
-                hash_t hash(0);
-                for (size_t j = 0; j < BITS; ++j) {
-                    if (v[j] > threshold) {
-                        hash |= (static_cast<hash_t>(1) << j);
-                    }
+        /* With bit counts appropriately tallied, create a 1 bit for each of
+         * the counts that's over threshold. That result is the hash. */
+        virtual hash_t result() /* override */ {
+            size_t threshold = hash_count / 2;
+            hash_t hash(0);
+            for (size_t j = 0; j < BITS; ++j) {
+                if (bit_counts[j] > threshold) {
+                    hash |= (static_cast<hash_t>(1) << j);
                 }
-                return hash;
             }
-        };
+            return hash;
+        }
+    };
+
+    class MovingWindowShingler: public HasherInterface {
+    private:
+        HasherInterface* hasher;
+        int window_size;
+        Cyclic<hash_t> cyclic;
+
     public:
-        typedef Hash      hash_type;
-
-        /* Alias of operator()
-         *
-         * Some languages have difficulty making use of operator(), and so this
-         * it made available for those languages */
-        inline hash_t hash(char **tokens) {
-            return operator()(tokens);
+        MovingWindowShingler(HasherInterface* hasher_, int window_size_)
```

Comment:
What's the intended ownership model for `hasher`? This should be documented in the arguments. It's currently unclear, and I see no attempts at cleaning them up. (All of the examples use locals.) It seems like maybe this should be a reference argument instead of a pointer argument.